Реализация задачи на блокирующей очереди.

DavidGelmanSerg opened the pull request.

DavidGelmanSerg committed:
Реализация задачи на блокирующей очереди.

Yuksovich commented on producer-consumer/src/main/java/ru/gelman/Main.java:
```diff
@@ -0,0 +1,41 @@
+            new Thread(produceTask).start();
+        }
+
+        Logger logger = LoggerFactory.getLogger(ConsumeTask.class);
```
это как тут оказалось? логгер у нас всегда private static final поле

DavidGelmanSerg replied:
исправил

Yuksovich commented on producer-consumer/src/main/java/ru/gelman/Main.java:
```diff
@@ -0,0 +1,41 @@
+
+        Logger logger = LoggerFactory.getLogger(ConsumeTask.class);
+        for (int i = 0; i < consumersAmount; i++) {
+            Consumer<Message> consumer = m -> logger.info("got " + m);
```
стоило уж сделать отдельную реализацию и там уже можно подобное воротить. Все-таки разделение на классы, принцип единой ответственности и прочие качества хорошего кода мы не отменяли

DavidGelmanSerg replied:
Исправил. Вынес Producer'a и Consumer'a в отдельные реализации.

Yuksovich commented on producer-consumer/src/main/java/ru/gelman/Message.java:
```diff
@@ -0,0 +1,10 @@
+
+public record Message(long id) {
+    @Override
+    public String toString() {
```
зачем переопределять toString у рекорда? он у него итак нормальный, почти такой же как ты написал

DavidGelmanSerg replied:
Исправил

Yuksovich commented on producer-consumer/src/main/java/ru/gelman/consumer/ConsumeTask.java:
```diff
@@ -0,0 +1,30 @@
+    private final Consumer<Message> consumer;
+    private final MessageStore store;
+    private final long consumeInterval;
+    public ConsumeTask(Consumer<Message> consumer, MessageStore store, long consumeInterval) {
```
отсутпы все-таки можно было сделать

DavidGelmanSerg replied:
Исправил

Yuksovich commented on producer-consumer/src/main/java/ru/gelman/producer/ProduceTask.java:
```diff
@@ -0,0 +1,35 @@
+    private final Supplier<Message> producer;
+    private final MessageStore store;
+    private final long produceInterval;
+    private final Logger logger = LoggerFactory.getLogger(ProduceTask.class);
```
static. А что с lombok? мы вроде обсуждали, что ты почитаешь про него и заиспользуешь

DavidGelmanSerg replied:
Исправил. Прикрутил lombok

Yuksovich commented on producer-consumer/src/main/java/ru/gelman/Main.java:
```diff
@@ -0,0 +1,41 @@
+
+    public static void main(String[] args) {
+        InputParser argsParser = new InputParser(args); //Взял этот класс из самой первой задачи: калькулятора
+        long consumersAmount = Long.parseLong(argsParser.getOptionValue(CONSUMERS_AMOUNT_KEY));
```
давай лучше задавать это в property файле, пора узнать, про такую штуку)

DavidGelmanSerg replied:
исправил

Yuksovich commented on producer-consumer/src/main/java/ru/gelman/consumer/ConsumeTask.java:
```diff
@@ -0,0 +1,30 @@
+
+import java.util.function.Consumer;
+
+public class ConsumeTask implements Runnable {
```
ConsumeTask и ProduceTask получились несколько непонятными сущностями. Т.е. максимально тупой продьюсер и консъюмер, при этом все делают эти таски. Непонятно зачем внедрены эти сущности, когда это должны были делать Producer и Consumer, а так получается несколько запутывает все

DavidGelmanSerg replied:
Исправил

Yuksovich commented on producer-consumer/src/main/java/ru/gelman/store/BlockingQueueMessageStore.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.concurrent.LinkedBlockingQueue;
+
+public class BlockingQueueMessageStore implements MessageStore {
+    private final BlockingQueue<Message> store;
```
в целом это правильное решение. Т.е. по сути лучшее что можно было придумать - использовать блокирующую очередь. Но давай попробуем сделать без применения такой штуки, а самим реализовать на примитивах синхронизации, используя synchronized.

DavidGelmanSerg replied:
Исправил

DavidGelmanSerg committed:
Реализация задачи на примитивах синхронизации - с использованием synchronized;
В проект добавлен Lombok;
Добавлена реализация потребителя;
Удалено разделение потребителя\производителя и задачи производства\потребления сообщений. Они объединены в единой реализации потребителя\производителя.

DavidGelmanSerg committed:
Удалены классы ConsumeTask и ProduceTask. Изменен оператор проверки заполненности очереди.

Yuksovich commented on producer-consumer/src/main/java/ru/gelman/Main.java:
```diff
@@ -1,41 +1,43 @@
+        Properties properties = getStartConfig();
+        long consumersAmount = Long.parseLong(properties.getProperty("consumer-amount"));
+        long producersAmount = Long.parseLong(properties.getProperty("producer-amount"));
+        long consumeInterval = Long.parseLong(properties.getProperty("consuming_delay"));
```
странно, что где-то дефис, где-то нижнее подчеркивание, стоит быть последовательным, иначе это запутывает

Yuksovich commented on producer-consumer/src/main/java/ru/gelman/Main.java:
```diff
@@ -1,41 +1,43 @@
-            ProduceTask produceTask = new ProduceTask(producer, store, produceInterval);
-            new Thread(produceTask).start();
+            var producer = new MessageProducer(store, produceInterval);
+            new Thread(producer).start();
```
почему бы не использовать ExecutorService. Это на будущее, ознакомься с этим интерфейсом и его имплементациями, иначе ты бесконтрольно создаешь потоки, и потом они бесконтрольно исполняются и завершить ты их никак не сможешь.

Yuksovich commented on producer-consumer/src/main/java/ru/gelman/consumer/MessageConsumer.java:
```diff
@@ -0,0 +1,34 @@
+
+@RequiredArgsConstructor
+public class MessageConsumer implements Consumer<Message>, Runnable {
+    private static final Logger logger = LoggerFactory.getLogger(MessageConsumer.class);
```
используя lombok, есть аннотация @Slf4j, чтобы не писать логгер руками

Yuksovich commented on producer-consumer/src/main/java/ru/gelman/Main.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    private static Properties getStartConfig() {
+        InputStream in = Thread.currentThread().getContextClassLoader().getResourceAsStream("start-config.properties");
```
не забывай закрывать ресурсы
